packages/core/src/renderAnnotations.ts

```diff
@@ -0,0 +1,186 @@
+import { Annotation, ImageAnnotation, TextAnnotation } from './types/Score'
+import { createElement } from './dom-utils'
+import classNames from 'classnames'
+
+export function renderAnnotations(
+  element: HTMLElement,
+  pregradingAnnotations: Annotation[],
+  censoringAnnotations: Annotation[]
+): void {
+  const annotations = [...pregradingAnnotations, ...censoringAnnotations]
+  const annotationsWithMessages = annotations.filter((a) => a.message)
+
+  for (const annotation of annotations) {
+    const type = pregradingAnnotations.includes(annotation) ? 'pregrading' : 'censoring'
+    const index = annotationsWithMessages.indexOf(annotation) + 1 || undefined
+
+    switch (annotation.type) {
+      case 'line':
+      case 'rect':
+        renderImageAnnotation(element, annotation, type, index)
+        break
+      default:
+        renderTextAnnotation(element, annotation, type, index)
+    }
+  }
+}
+
+function renderImageAnnotation(
+  element: HTMLElement,
+  annotation: ImageAnnotation,
+  type: 'pregrading' | 'censoring',
+  index?: number
+): void {
+  const { attachmentIndex } = annotation
+  const image = element.querySelectorAll('img')[attachmentIndex]
+
+  const wrapper = getWrapper()
+  const shape = mkShape()
+  wrapper.appendChild(shape)
+
+  function getWrapper() {
+    const parent = image.parentElement!
+
+    if (parent instanceof HTMLSpanElement) {
+      return parent
+    }
+
+    const wrapper = createElement('span', { className: 'e-annotation-wrapper' })
+    parent.insertBefore(wrapper, image)
+    wrapper.appendChild(image)
+
+    return wrapper
+  }
+
+  function mkShape() {
+    const shape = createElement('mark', {
+      className: classNames('e-annotation e-annotation--shape', {
+        'e-annotation--pregrading': type === 'pregrading',
+        'e-annotation--censoring': type === 'censoring',
+      }),
+      title: annotation.message,
+    })
+
+    const style = shape.style
+    const pct = (n: number) => `${n * 100}%`
+
+    if (annotation.type === 'rect') {
+      style.left = pct(annotation.x)
```

Comment:
jos tykkää kikkailla, niin näin voi myös sanoa:
```
Object.assign(style, { left: ..., top: ..., right: ..., bottom: ...})
```

jolloin noi neljä kenttää vois laittaa jounku funktion palauttamaks ja tehdä koodissa ton assignin vaan kerran, tyyliin:
```
Object.assign(style, getDimensions(annotation))
```
ja getDimensions vois käyttää object spreadia noille arvoille